File: Tree/connect-nodes-at-same-level.cpp
//https://practice.geeksforgeeks.org/problems/connect-nodes-at-same-level/1

#include <bits/stdc++.h>
using namespace std;
struct Node
{
  int data;
  struct Node *left;
  struct Node *right;
  struct Node *nextRight;
};
void connect(struct Node *p);
/* Helper function that allocates a new node with the
   given data and NULL left and right pointers. */
struct Node* newNode(int data)
{
  struct Node* node = new Node;
  node->data = data;
  node->left = NULL;
  node->right = NULL;
  return(node);
}
void printSpecial(Node *root)
{
   if (root == NULL)
     return;
   Node *node = root;
   while (node != NULL)
   {
      printf("%d ", node->data);
      node = node->nextRight;
   }
   if (root->left)
     printSpecial(root->left);
   else
     printSpecial(root->right);
}
void inorder(Node *root)
{
    if (root == NULL)
       return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}
/* Driver program to test size function*/
int main()
{
  int t;
  scanf("%d
", &t);
  while (t--)
  {
     map<int, Node*> m;
     int n;
     scanf("%d",&n);
     struct Node *root = NULL;
     struct Node *child;
     while (n--)
     {
        Node *parent;
        char lr;
        int n1, n2;
        scanf("%d %d %c", &n1, &n2, &lr);
        if (m.find(n1) == m.end())
        {
           parent = newNode(n1);
           m[n1] = parent;
           if (root == NULL)
             root = parent;
        }
        else
           parent = m[n1];
        child = newNode(n2);
        if (lr == 'L')
          parent->left = child;
        else
          parent->right = child;
        m[n2]  = child;
     }
     connect(root);
     printSpecial(root);
     printf("
");
     inorder(root);
     printf("
");
  }
  return 0;
}


/*Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function is mentioned above.*/

/* struct Node
{
  int data;
  Node *left,  *right;
  Node *nextRight;  // This has garbage value in input trees
}; */
// Should set the nextRight for all nodes
void connect(Node *p)
{
   // Your Code Here
   if(root==NULL) return;
   queue <Node *>q;
   q.push(root);
   
   while(q.size()>0)
   {
       int size=q.size();
       
       while(size>0)
       {
           Node *front=q.front();
           q.pop();
            
            if(front->left) q.push(front->left);
            if(front->right)q.push(front->right);
           
           if(size==1)
            front->nextRight=NULL;
            else
            {
                front->nextRight=q.front();
            }
           size--;
       }
   }
   
}

File: Tree/connect-nodes-at-same-level.py
"""
https://www.geeksforgeeks.org/connect-nodes-at-same-level/
"""

# Recursive Python program for level 
# order traversal of Binary Tree
# Method 2 (Extend Pre Order Traversal) 


# A node structure
from operator import le


class Node:

	# A utility function to create a new node
	def __init__(self, key):
		self.data = key 
		self.left = None
		self.right = None
		self.nextright = None

			
def LevelOrder(root):
    if not root:
        return None
    q=[]
    q.append(root)
    while q:
        prv=Node(None)
        for _ in range(len(q)):
            tq=q.pop(0)
            if tq.left:
                q.append(tq.left)
            if tq.right:
                q.append(tq.right)
            if prv is not None:
                prv.nextright=tq
            prv=tq
        prv.nextright=None
        
        

# Driver program to test above function
root = Node(3)
root.left = Node(9)
root.right = Node(20)
# root.right.left = Node(15)
# root.right.right = Node(7)

print("Level order traversal of binary tree is -")
print(LevelOrder(root))
print("Next right of {} is {}".format(root.left.data, root.left.nextright.data))
# print("Next right of {} is {}".format(root.right.data, root.right.nextright.data))